### PR TITLE
Issue 4428 - Federation's containers have group

### LINKED ISSUE #4428: Update the frontend to understand the new federation settings/ federation stats schema
### Description
We need to update the front end to understand the new schema, which is being returned by federation settings and federation stats.
We also need to update the edit federation panel to allow users to define groupings for their federations.

See scoping document for details: https://github.com/3drepo/3D-Repo-Product-Team/issues/304

#### READ ME
- Please branch from https://github.com/3drepo/3drepo.io/issues/4391
- This will also require https://github.com/3drepo/3drepobouncer/issues/639 for the full experience (though not necessary for the frontend work)

### Goals
- [x] The user should be able to view federations as before (currently broken on v5 viewer due to it not understanding the new containers array)
- [x] The user should be able to edit federations and optionally add grouping should they wish to
  - [x] existing groups should appear as a suggestion to minimise chance of typo errors

### Tasks
- [ ] Adapt the `IFederation` interface to work with the new `containers` prop;
- [ ] Use new edit federation design (do not worry about the new modal design just yet);
- [ ] Add a autosuggest in the included containers to manipulate the group

### Related Resources
Scoping document: https://github.com/3drepo/3D-Repo-Product-Team/issues/304
Federation edit mockup: https://www.figma.com/file/bRqIl0g9psqD3ISkmxwcNp/3D-Repo?type=design&node-id=605%3A38114&mode=design&t=1PFl8RzEZtGnLqX2-1

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #4428

#### Description
- The frontend understands the new federation interface and adheres to it
- The frontend also uses the newly created design for federation settings (not the modal UI, an issue has been created to address that as it's something generic https://github.com/3drepo/3drepo.io/issues/4435)

#### Test cases
- Use the edit/create federation as before to add/remove containers (regression test);
- Assign groups to the containers (only groups in the federation should allow groups)

### CODE REVIEW

**Comment by george-hadfield** [comment]
### Add right padding to plus and minus icons

Just needs a bit more to match the design

image

**Comment by george-hadfield** [comment]
### Change plus icon colour when on blue background
Change to white

image

**Review by george-hadfield — CHANGES_REQUESTED** [review]
Couple of small ones

**Review by george-hadfield — APPROVED** [review]

**Comment by carmenfan** [comment]
@Amantini1997 i merged in changes from the parent branch and it had conflicts (https://github.com/3drepo/3drepo.io/pull/4436/commits/ffd41431179e56ff56699159d7d6308095647efc)

and it now white screens 😢

**Comment by Amantini1997** [comment]
> @Amantini1997 i merged in changes from the parent branch and it had conflicts ([ffd4143](https://github.com/3drepo/3drepo.io/commit/ffd41431179e56ff56699159d7d6308095647efc))
> 
> and it now white screens 😢

It should work again now, although the "cancel" button in the modals 

> @Amantini1997 i merged in changes from the parent branch and it had conflicts ([ffd4143](https://github.com/3drepo/3drepo.io/commit/ffd41431179e56ff56699159d7d6308095647efc))
> 
> and it now white screens 😢

It should work again now

File: frontend/src/v5/ui/routes/dashboard/projects/federations/editFederationModal/editFederationContext.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		const container: GroupedContainer = { _id };
+		const group = groupsByContainer[_id];
+		if (group) {
+			container.group = group;
+		}
+		return container;
+	});
+
+	const existingGroups = federations.flatMap((f) => f.containers.map(({ group }) => group));
+	const groups = uniq(existingGroups.concat(Object.values(groupsByContainer))).filter(Boolean);
```
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/federations/editFederationModal/editFederationContext.tsx:64]: image
These should probably be sorted alphabetically

File: frontend/src/v5/ui/routes/dashboard/projects/federations/editFederationModal/editFederation/editFederationContainersList/editFederationContainersListItem/editFederationContainersListItem.styles.ts [hunk 2]
```diff
@@ -28,3 +29,21 @@ export const DashboardListItemRow = styled(DashboardListItemRowBase)`
 		margin-right: 20px;
 	}
 `;
+
+export const Autocomplete = styled(AutocompleteBase)`
```
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/federations/editFederationModal/editFederation/editFederationContainersList/editFederationContainersListItem/editFederationContainersListItem.styles.ts:33]: image
There's a shadow floating when the list is empty

**Review by The-Daniel — CHANGES_REQUESTED** [review]

**Review by carmenfan — CHANGES_REQUESTED** [review]
Can we make a change on the groups suggested list drop down?

Currently if I type something that doesn't exist and I tab away it gets deleted. This is a bit of an expected behaviour (and it's quite annoying!). Can we assume that if the user change focus, they want to keep that group name instead of a cancel behaviour?

img-22-40-23-04-40


If this changes the behaviour in the upload revision modal that's fine - that one is also really annoying.

Essentially I think we'd like the behaviour to be closer to v4 risks

img-22-42-23-04-42

**Comment by Amantini1997** [comment]
@carmenfan I only changed the behvaiour of the autocomplete in the edit/create federation modal.
I tried to tackle the upload file modal as well, but that impacted performance badly and since Dan is already refactoring that code, I decided to wait until then to do it again

**Review by The-Daniel — APPROVED** [review]

**Review by carmenfan — APPROVED** [review]

**Pull request merged**